tests/amethyst_test/src/amethyst_application.rs

```diff
@@ -0,0 +1,1287 @@
+//   ergonomics of this test harness.
+type FnResourceAdd = Box<FnMut(&mut World) + Send>;
+type FnState<T, E> = SendBoxFnOnce<'static, (), Box<State<T, E>>>;
+
+type DefaultPipeline = PipelineBuilder<
+    Queue<(
+        Queue<()>,
+        StageBuilder<Queue<(Queue<(Queue<()>, DrawSprite)>, DrawUi)>>,
+    )>,
+>;
+
+/// Screen width used in predefined display configuration.
+pub const SCREEN_WIDTH: u32 = 800;
+/// Screen height used in predefined display configuration.
+pub const SCREEN_HEIGHT: u32 = 600;
+/// The ratio between the backing framebuffer resolution and the window size in screen pixels.
+/// This is typically one for a normal display and two for a retina display.
+pub const HIDPI: f64 = 1.;
+
+// Use a mutex to prevent multiple tests that open GL windows from running simultaneously, due to
+// race conditions causing failures in X.
+// <https://github.com/tomaka/glutin/issues/1038>
+lazy_static! {
+    static ref X11_GL_MUTEX: Mutex<()> = Mutex::new(());
+}
+
+/// Builder for an Amethyst application.
+///
+/// This provides varying levels of setup so that users do not have to register common bundles.
+///
+/// # Type Parameters
+///
+/// * `T`: Game data type that holds the common dispatcher.
+/// * `E`: Custom event type shared between states.
+#[derive(Derivative, Default)]
+#[derivative(Debug)]
+pub struct AmethystApplication<T, E>
+where
+    E: Send + Sync + 'static,
+{
+    /// Functions to add bundles to the game data.
+    ///
+    /// This is necessary because `System`s are not `Send`, and so we cannot send `GameDataBuilder`
+    /// across a thread boundary, necessary to run the `Application` in a sub thread to avoid a
+    /// segfault caused by mesa and the software GL renderer.
+    #[derivative(Debug = "ignore")]
+    bundle_add_fns: Vec<BundleAddFn>,
+    /// Functions to add bundles to the game data.
+    ///
+    /// This is necessary because `System`s are not `Send`, and so we cannot send `GameDataBuilder`
+    /// across a thread boundary, necessary to run the `Application` in a sub thread to avoid a
+    /// segfault caused by mesa and the software GL renderer.
+    #[derivative(Debug = "ignore")]
+    resource_add_fns: Vec<FnResourceAdd>,
+    /// States to run, in user specified order.
+    #[derivative(Debug = "ignore")]
+    state_fns: Vec<FnState<T, E>>,
+    /// Game data and event type.
+    state_data: PhantomData<(T, E)>,
+    /// Whether or not this application uses the `RenderBundle`.
+    render: bool,
+}
+
+impl AmethystApplication<GameData<'static, 'static>, ()> {
+    /// Returns an Amethyst application without any bundles.
+    pub fn blank() -> AmethystApplication<GameData<'static, 'static>, ()> {
+        AmethystApplication {
+            bundle_add_fns: Vec::new(),
+            resource_add_fns: Vec::new(),
+            state_fns: Vec::new(),
+            state_data: PhantomData,
+            render: false,
+        } // kcov-ignore
```

Comment:
Whats that?

Reply:
remnant of my code coverage testing, i can remove them